lib/src/iterable_matchers.dart

```diff
@@ -206,6 +216,31 @@ class _UnorderedMatches extends Matcher {
   Description describeMismatch(item, Description mismatchDescription,
           Map matchState, bool verbose) =>
       mismatchDescription.add(_test(item));
+
+  /// Returns [true] if the value at [valueIndex] can be paired with some
+  /// unmatched matcher and updates the state of [matched].
+  ///
+  /// If there is a conflic where multiple values may match the same matcher
+  /// recursively looks for a new place to match the old value. [seen] trackes
+  /// the matchers that have been used _during_ this search.
+  bool _findPairing(List<List<int>> edges, int valueIndex, List<int> matched,
+      [Set<int> seen]) {
+    seen ??= new Set<int>();
+    final possiblePairings = edges[valueIndex].where((m) => !seen.contains(m));
```

Comment:
I was going to say this doesn't allow you to bail early, but then I remembered Iterables are lazy :D. Dart ftw!

Reply:
😄